drush/civicrm.drush.inc

```diff
@@ -383,15 +383,25 @@ function _civicrm_install_db($dbuser, $dbpass, $dbhost, $dbname,
   $siteRoot = drush_get_context('DRUSH_DRUPAL_SITE_ROOT', FALSE);
 
   $sqlPath = "$modPath/civicrm/sql";
-  $conn = @mysqli_connect($dbhost, $dbuser, $dbpass);
-  if (!@mysqli_select_db($conn, $dbname) &&
-    !@mysqli_query($conn, "CREATE DATABASE $dbname")
-  ) {
+
+  if (function_exists('mysqli_connect')) {
+    $dbhostParts = explode(':', $dbhost);
+    $conn = @mysqli_connect($dbhostParts[0], $dbuser, $dbpass, '', isset($dbhostParts[1]) ? $dbhostParts[1] : '');
+    $dbOK = ($a = @mysqli_select_db($conn, $dbname)) || ($b = @mysqli_query($conn, "CREATE DATABASE $dbname"));
+  }
+  elseif (function_exists('mysql_connect')) {
+    $conn = @mysql_connect($dbhost, $dbuser, $dbpass);
+    $dbOK = @mysql_select_db($dbname, $conn) || @mysql_query("CREATE DATABASE $dbname", $conn);
+  }
+  else {
+    $dbOK = FALSE;
+  }
+  if (!$dbOK) {
     drush_die(dt('CiviCRM database was not found. Failed to create one.'));
   }
 
   // setup database with civicrm structure and data
-  $dsn = "mysqli://{$dbuser}:{$dbpass}@{$dbhost}/{$dbname}?new_link=true";
+  $dsn = "mysql://{$dbuser}:{$dbpass}@{$dbhost}/{$dbname}?new_link=true";
```

Comment:
@seamuslee001, yes, but this is not the right place to handle that.

Look at this way:
 * There's the *normal* case where we connect to MySQL through the PEAR DB abstractions (e.g. `DB::connect()`). If there's a DSN like `mysql://user:pass@host/name`, then we're using PEAR DB abstractions. We want to cleanly switch all of these from `mysql://` to `mysqli://` in one go -- which only requires a 1-line patch to `packages/DB.php`. It's much harder to hunt down all the things which refer to `mysql://` and change them one-by-one.
 * There are also the evil cases where someone circumvented PEAR DB and connected directly with a lower-level API. The main place we see this is installation code (like lines 386-390). (*Theory*: Circumventing PEAR DB allows precise error-handling in edge-cases like "the user is valid but the database has not been created yet".) 

If the code constructs a DSN like `mysql://`, stick with `mysql://`, then keep it that way. If the code circumvents PEAR DB (e.g. `mysql_connect()`), then it needs a patch. (An adaptive approach which picks between `mysql_*()` and `mysqli_*()` is nice because we don't have to think as hard when juggling PRs across different repo's, but it's not strongly required. The official policy will be that only one driver is supported.)